### PR TITLE
expression: only do a shallow copy when evaluating a "Column" expression

### PR DESCRIPTION
**Pull request by zz-jason**
1. implement an `evaluatorSuit` to handle the expression evaluation.
2. for column expressions, we only do a shallow copy instead of deep copy.

For query `SELECT L_QUANTITY AS a, L_QUANTITY AS b, L_QUANTITY AS c FROM lineitem` on tpch `lineitem` table(6001215 rows), the execution time can be reduced from 10.90s to 7.11s

### CODE REVIEW

**Comment by ngaut** [comment]
Good job.

File: expression/evaluator.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+		for i, length := 1, len(outputIdxes); i < length; i++ {
+			output.MakeRef(outputIdxes[0], outputIdxes[i])
+		}
+	}
+	return nil
+}
+
+type otherEvaluator struct {
+	outputIdxes  []int
+	scalaFunc    []Expression
```
Comment (coocood) [inline expression/evaluator.go]: scalarFuncs

File: expression/evaluator.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+	for inputIdx, outputIdxes := range e.inputIdx2OutputIdxes {
+		output.SwapColumn(outputIdxes[0], input, inputIdx)
+		for i, length := 1, len(outputIdxes); i < length; i++ {
+			output.MakeRef(outputIdxes[0], outputIdxes[i])
+		}
+	}
+	return nil
+}
+
+type otherEvaluator struct {
```
Comment (coocood) [inline expression/evaluator.go]: defaultEvaluator is better?

File: expression/evaluator.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+		}
+	}
+	return nil
+}
+
+// EvaluatorSuit is responsible for the evaluation of a list of expressions.
+// It seperates them to "column" and "other" expressions and evaluates "other"
+// expressions before "column" expressions.
+type EvaluatorSuit struct {
+	evaluator4Col   *columnEvaluator
```
Comment (coocood) [inline expression/evaluator.go]: Why not just `columnEvaluator`?

File: expression/evaluator.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+	evaluator4Other *otherEvaluator
+}
+
+// NewEvaluatorSuit creates an EvaluatorSuit to evaluate all the exprs.
+func NewEvaluatorSuit(exprs []Expression) *EvaluatorSuit {
+	e := &EvaluatorSuit{
+		evaluator4Col: &columnEvaluator{
```
Comment (coocood) [inline expression/evaluator.go]: How about lazy initialize?
There may be no column expressions or all column expression.

**Comment by shenli** [comment]
@hanfei1991 PTAL

File: expression/evaluator.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+package expression
+
+import (
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+type columnEvaluator struct {
+	inputIdx2OutputIdxes map[int][]int
```
Comment (coocood) [inline expression/evaluator.go]: `2O` can be easily misread as `20`, I think `inputIdxToOutputIdxes` is better.

**Comment by coocood** [comment]
LGTM

**Pull request merged**